[Documentation] Add document describing Cart/Checkout data flow (#53512)
woocommerce/woocommerce: docs/cart-and-checkout-blocks/overview-of-data-flow.md

---
post_title: Cart and Checkout - Overview of data flow between client and server
menu_title: Data flow
tags: reference
---

# WooCommerce Cart/Checkout Data Flow Overview

## Overview

> 

Reviewer: Might be good to start by expanding a bit on what `data` is here and give some examples.

> In the WooCommerce Cart and Checkout blocks, the server is the source of truth, and data should be persisted server side. Things like component state (e.g. the checkout order notes, or coupon form input data) should not be persisted on the server, unless it's important that this state is persisted across page loads.

Reviewer: `component state` - this is quite vague, I'm not sure what you mean here? Local state in a react component? Data stores? Context? State is stored at many levels.

Also I'm working on a PR that will persist additional fields, order notes and selected payment methods. So actually we're heading in the opposite direction of persisting everything to the server - which makes sense as it's the source of truth.

Author: > Also I'm working on a PR that will persist additional fields, order notes and selected payment methods. So actually we're heading in the opposite direction of persisting everything to the server - which makes sense as it's the source of truth.

Great! I think my point is still OK here though, selected payment method could be considered important to persist.

You may wish to get data from the server into the client, and vice/versa. This document will outline the general concepts and data flow in the Cart/Checkout blocks, and provide links, or guidance on some common use cases.

### Where is data stored?

> All the data relating to a cart, customer, and order is stored on the server, either in the database or in the customer's session. When it is sent to the server, it is stored in [`@wordpress/data`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/) data stores.

Reviewer: ```suggestion
All the data relating to a cart, customer, and order is stored on the server, either in the database or in the customer's session. When it is sent to the client, it is stored in [`@wordpress/data`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/) data stores.
```

### How do I get my Server-side (PHP) data into the client (JavaScript) and vice/versa

Since the server is the source of truth, all data from the client should eventually make its way there if it needs to be persisted with the order.

There is also the case that your client needs data that can only be derived server-side, such as configuration options set in the WooCommerce settings dashboard, or data coming from external services (e.g. courier rates or payment APIs).

#### Server (PHP) → Client (JavaScript)

> There are different ways to get your data from the server to the client on whether your it is static (e.g. an option from the WooCommerce settings dashboard) or if it is dynamic/might change during the course of a shopper's session (e.g. tax rates or fees based on the shipping address).

Reviewer: ```suggestion
There are two ways to get your data from the server to the client, depending on wether it is static or  dynamic
```

##### Static data

> If the data is static, i.e. it is not likely to change based on actions the shopper takes then a good way of getting this data to the client might be through the `AssetDataRegistry`.

Reviewer: ```suggestion
Static data is not likely to change based on actions the shopper takes. For example, an option from the WooCommerce settings dashboard. A good way of getting this data to the client might be through the `AssetDataRegistry`.
```

When data is added here, it is serialized and sent to the client on page load. It will not change based on actions the shopper takes, e.g. adding items to the cart or changing their address.

A key/value pair can be added to the registry like so:

```php
add_action(
    'woocommerce_blocks_loaded',
    function() {
      $asset_data_registry = \Automattic\WooCommerce\Blocks\Package::container()->get( \Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry::class );
      $asset_data_registry->add( 'namespace/value', 'this is a custom value' );
    }
);
```

> If a duplicate key exists, it will not be overwritten. Using a unique identifier is important.

Reviewer: Maybe mention the format `namespace/value` in order to ensure uniqueness

To get this data on the client, use `wc.wcSettings.getSetting` like so:

```js
const myCustomValue = wc.wcSettings.getSetting( 'namespace/value', 'Fallback value.' );
```

#### Dynamic data

> If data might change, e.g. based on a shopper's location or items in the cart, it is a good idea to add this to the cart API response. The cart response is sent on many routes during the shopper's journey, and on almost every API response triggered by the Cart/Checkout blocks.

Reviewer: ```suggestion
Dynamic data is data that is likely to change in response to the shopper's actions. For example, changing location, or items in the cart. In this case, you need to add this to the cart API response. The cart response is sent on many routes during the shopper's journey, and on almost every API response triggered by the Cart/Checkout blocks.
```

> To add data here, you'll need to extend the API response. The [Exposing your data in the Store API](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-data.md)

Reviewer: ```suggestion
To add data here, you'll need to extend the API response. See [Exposing your data in the Store API](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-data.md)
```

> document outlines how to do this.

Reviewer: ```suggestion
```

#### Client (JavaScript) → Server (PHP)

Getting data from the client to the server can be done in a couple of different ways too. The options are: piggyback on a Store API request when one is made, or send the data on demand.

##### Piggybacking on a Store API request

> This is useful for things that don't require an immediate response from the server, for example if you've added a new block to the Checkout and it contains a form field that should be saved along with the order. If this form field just needs to be saved and doesn't need to update any other values in the cart, then sending it with the checkout request is a good idea. The [Add a new inner block containing a custom field to the WooCommerce Checkout Block](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/rest-api/extend-rest-api-add-custom-fields.md) documentation provides a worked example of using `setExtensionData` along with extending the Store API to receive your data in an existing request.

Reviewer: The `PUT` endpoint on checkout will likely make this easier, but we can update the docs at a later date

#### Entering customer data into the checkout form

When the shopper enters data into the form, the data is immediately written to the `wc/store/cart` data store, and a debounced method called [`pushChanges`](https://github.com/woocommerce/woocommerce/blob/4861ec250ef1789f814f4209755165e8abe7b838/plugins/woocommerce-blocks/assets/js/data/cart/push-changes.ts#L167) is called. This method sends the customer data to the server where it is persisted against the customer. The full cart is sent back as a response, and the data store is updated with this. 

This is important to note, because if any code is running on the server that modifies the customer addresses, then it will be reflected in the response.

For example, if a plugin modifies the address data to ensure all city names are capitalised, and the shopper enters "london" into the city, when the data is returned to the client, the text would change to "London" and the input field would update.

> Modifying form fields while the shopper is interacting with them _could_ be a jarring experience.

Reviewer: If we mention this, do you think we should also say how to mitigate against it?

Author: Yeah, good idea. I added

> Instead of making these changes while the user is interacting with the form, consider making them while processing the checkout action on the server.

Thanks
